src/WasmModule.ts

```diff
@@ -0,0 +1,193 @@
+import { decode } from '@webassemblyjs/wasm-parser';
+import ast from '@webassemblyjs/ast';
+
+import Graph from './Graph';
+import { IdMap, ModuleJSON, WebAssemblyJSAst } from './rollup/types';
+
+import { makeLegal } from './utils/identifierHelpers';
+import { RenderOptions } from './utils/renderHelpers';
+import { basename, extname } from './utils/path';
+
+import ExternalModule from './ExternalModule';
+import Module from './Module';
+
+import ModuleScope from './ast/scopes/ModuleScope';
+import Variable from './ast/variables/Variable';
+import NamespaceVariable from './ast/variables/NamespaceVariable';
+
+const decoderOpts = {
+	ignoreCodeSection: true,
+	ignoreDataSection: true
+};
+
+const buildLoader = ({ URL, IMPORT_OBJECT }) => `
+	function then(resolve) {
+		if (typeof WebAssembly.instantiateStreaming !== 'function') {
+		  throw new Error('WebAssembly.instantiateStreaming is not supported');
+		}
+
+		if (typeof window.fetch !== 'function') {
+		  throw new Error('window.fetch is not supported');
+		}
+
+		const req = window.fetch('${URL}');
+
+		WebAssembly
+			.instantiateStreaming(req, ${IMPORT_OBJECT || '{}'})
+			.then(res => res.instance.exports)
+			.then(resolve)
+			.catch(resolve);
+	}
+`;
+
+export interface ExportDescription {
+	localName: string;
+	identifier?: string;
+}
+
+export interface ImportDescription {
+	source: string;
+	name: string;
+	module: Module | ExternalModule | null;
+}
+
+export default class WasmModule {
+	id: string;
+	graph: Graph;
+	code: Buffer;
+
+	scope: ModuleScope;
+
+	ast: WebAssemblyJSAst.Program;
+
+	sources: string[];
+	resolvedIds: IdMap;
+
+	dynamicImportResolutions: {
+		alias: string;
+		resolution: Module | ExternalModule | string | void;
+	}[];
+	dependencies: (Module | ExternalModule | WasmModule)[];
+
+	imports: { [name: string]: ImportDescription };
+	exports: { [name: string]: ExportDescription };
+
+	// this is unused on Module,
+	// only used for namespace and then ExternalExport.declarations
+	declarations: {
+		'*'?: NamespaceVariable;
+		[name: string]: Variable | undefined;
+	};
+
+	isExternal: false;
+
+	constructor(graph: Graph, id: string) {
+		this.id = id;
+		this.graph = graph;
+		this.code = new Buffer('');
+
+		this.ast = null;
+
+		// imports and exports, indexed by local name
+		this.imports = Object.create(null);
+		this.exports = Object.create(null);
+		this.resolvedIds = Object.create(null);
+		this.declarations = Object.create(null);
+
+		// all dependencies
+		this.dynamicImportResolutions = [];
+		this.sources = [];
+		this.dependencies = [];
+
+		this.scope = new ModuleScope(this);
+
+		// expose Thenable
+		this.exports.then = {
+			localName: 'then'
+		};
+	}
+
+	render(options: RenderOptions) {
+		const URL = 'foo';
+		const content = buildLoader({ URL });
+
+		return { trim() {}, content };
```

Comment:
I'm not sure how to constructor the MagicString object in that case.

Reply:
WasmModule.render must return a Buffer surely? I think we should handle the distinction in the caller sites.

Reply:
I see. Following my comment https://github.com/rollup/rollup/pull/2113#discussion_r179798411, since i'm writing the file myself I bypass this logic actually, but I use it to inject the loader needed on the JS module. Does it make sense to you?